### PR TITLE
txn: check conflicts for pessimistic prewrite according to for_update_ts

### LINKED ISSUE #11612: Sysbench read and write workload hit 9007 Write conflict error
### What version of TiKV are you using?

/ # /tikv-server -V
TiKV
Release Version:   5.3.0
Edition:           Community
Git Commit Hash:   6c1424706f3d5885faa668233f34c9f178302f36
Git Commit Branch: heads/refs/tags/v5.3.0
UTC Build Time:    2021-11-22 10:51:38
Rust Version:      rustc 1.56.0-nightly (2faabf579 2021-07-27)
Enable Features:   jemalloc mem-profiling portable sse protobuf-codec test-engines-rocksdb cloud-aws cloud-gcp
Profile:           dist_release

### Steps to reproduce

run oltp test case oltp_stability_002 failed.

### What did you expect?
Case run succeeded.

### What did happened?
Hit following error.
2021-12-08T12:03:47.527+0800	INFO	oltppkg/workload.go:331	FATAL: mysql_stmt_execute() returned error 9007 (Write conflict, txnStartTS=429637298355962059, conflictStartTS=429637298355962016, conflictCommitTS=429637298382176376, key={tableID=156, indexID=1, indexValues={5025659, 5045579, }} primary={tableID=156, handle=5045579} [try again later]) for query 'COMMIT'

/var/lib/tidb/log # grep -A5 -B 5 '429637298355962016' *.log
tidb.log-[2021/12/08 04:03:47.090 +00:00] [WARN] [pd.go:152] ["get timestamp too slow"] ["cost time"=31.14879ms]
tidb.log-[2021/12/08 04:03:47.265 +00:00] [ERROR] [terror.go:307] ["encountered error"] [error=EOF] [stack="github.com/pingcap/tidb/parser/terror.Log\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/parser/terror/terror.go:307\ngithub.com/pingcap/tidb/server.(*Server).onConn\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/server/server.go:519"]
tidb.log-[2021/12/08 04:03:47.523 +00:00] [WARN] [client_batch.go:794] ["wait response is cancelled"] [to=tc-tikv-2.tc-tikv-peer.testbed-oltp-hm-k52vt.svc:20160] [cause="context canceled"]
tidb.log-[2021/12/08 04:03:47.524 +00:00] [WARN] [client_batch.go:794] ["wait response is cancelled"] [to=tc-tikv-1.tc-tikv-peer.testbed-oltp-hm-k52vt.svc:20160] [cause="context canceled"]
tidb.log-[2021/12/08 04:03:47.524 +00:00] [WARN] [client_batch.go:794] ["wait response is cancelled"] [to=tc-tikv-3.tc-tikv-peer.testbed-oltp-hm-k52vt.svc:20160] [cause="context canceled"]
tidb.log:[2021/12/08 04:03:47.524 +00:00] [WARN] [session.go:721] ["can not retry txn"] [conn=29083] [label=general] [error="[kv:9007]Write conflict, txnStartTS=429637298355962059, conflictStartTS=429637298355962016, conflictCommitTS=429637298382176376, key={tableID=156, indexID=1, indexValues={5025659, 5045579, }} primary={tableID=156, handle=5045579} [try again later]"] [IsBatchInsert=false] [IsPessimistic=true] [InRestrictedSQL=false] [tidb_retry_limit=10] [tidb_disable_txn_auto_retry=true]
tidb.log:[2021/12/08 04:03:47.524 +00:00] [WARN] [session.go:737] ["commit failed"] [conn=29083] ["finished txn"="Txn{state=invalid}"] [error="[kv:9007]Write conflict, txnStartTS=429637298355962059, conflictStartTS=429637298355962016, conflictCommitTS=429637298382176376, key={tableID=156, indexID=1, indexValues={5025659, 5045579, }} primary={tableID=156, handle=5045579} [try again later]"]
tidb.log:[2021/12/08 04:03:47.524 +00:00] [INFO] [conn.go:1069] ["command dispatched failed"] [conn=29083] [connInfo="id:29083, addr:10.233.116.131:46984 status:10, collation:utf8mb4_general_ci, user:root"] [command=Execute] [status="inTxn:0, autocommit:1"] [sql=COMMIT] [txn_mode=PESSIMISTIC] [err="[kv:9007]Write conflict, txnStartTS=429637298355962059, conflictStartTS=429637298355962016, conflictCommitTS=429637298382176376, key={tableID=156, indexID=1, indexValues={5025659, 5045579, }} primary={tableID=156, handle=5045579} [try again later]\nCOMMIT"]

Logs and monitor can be get from minio via testbed testbed-oltp-hm-k52vt.

### PR DESCRIPTION
**Pull request by sticnarf**
Signed-off-by: Yilin Chen <sticnarf@gmail.com>

### What is changed and how it works?

Issue Number: Close #11612

What's Changed:

```commit-message
For CDC compatibility, we check conflicts for retried pessimistic
prewrite (#11264). But this may cause false WriteConflict errors
if the commit TS of the record is larger than the start ts but smaller
than the for_update_ts.

So, this commit uses for_update_ts to check conflicts for pessimistic
transactions to avoid returning false errors. And if there really is a
conflicted record, PessimisticLockNotFound is returned to become
more semantically correct.
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- 
### Release note 

```release-note
Prevent false WriteConflict errors in pessimistic transactions.
```

### CODE REVIEW

File: src/storage/txn/actions/prewrite.rs [hunk 1]
```diff
@@ -1417,18 +1442,40 @@ pub mod tests {
...
             kvproto::kvrpcpb::Assertion::None,
             kvproto::kvrpcpb::AssertionLevel::Off,
         );
-        must_locked(&engine, b"k2", 10);
+        must_locked(&engine, b"k2", 12);
+        must_rollback(&engine, b"k2", 12, false);
+
+        // And conflict check is according to the for_update_ts for pessimistic prewrite.
+        // So, it will not report error if for_update_ts is large enough.
+        must_prewrite_put_impl(
```
Comment (cfzjywxk) [inline src/storage/txn/actions/prewrite.rs:1461]: Maybe we should also add some test cases in the `utf` suite to cover the path that pessimistic prewrite happens on index keys, it's a bit complex to tell when to do the check and what's the expected output by now considering async commit protocol and assertion check requirements.
/cc @MyonKeminta @ekexium @zyguan 
What do you think?

**Review by cfzjywxk — APPROVED** [review]

**Comment by MyonKeminta** [comment]
Did you confirm how the `for_update_ts` field is passed from the client? Different statements in the same transaction may execute with different transactions, so the keys may be written with different `for_update_ts`. If in prewrite phase the keys are checked with different `for_update_ts`, I'm afraid there might be other problems...

**Comment by sticnarf** [comment]
> Did you confirm how the `for_update_ts` field is passed from the client? Different statements in the same transaction may execute with different transactions, so the keys may be written with different `for_update_ts`. If in prewrite phase the keys are checked with different `for_update_ts`, I'm afraid there might be other problems...

I think all prewrite requests contain the final `for_update_ts`, which is the biggest one.

Some keys may be locked with a smaller `for_update_ts`. And that's why `amend_pessimistic_lock` can only use `start_ts` instead of the final `for_update_ts`.

But I think it won't cause a problem here. What we want through the extra constraint check is preventing prewriting when the same transaction has been committed. In this case, the `commit_ts` must be greater than the final `for_update_ts`.

**Review by MyonKeminta — APPROVED** [review]

File: src/storage/txn/actions/prewrite.rs (hunk 1 above)
Comment (ekexium) [reply src/storage/txn/actions/prewrite.rs:1461]: What's the purpose of the test?

**Review by youjiali1995 — APPROVED** [review]

**Pull request merged**